grafana/grafana
# Elasticsearch: Implement processing of raw document query results in backend

Part of https://github.com/grafana/grafana/issues/54011. This PR adds processing of raw document response to backend. To do this, we have encapsulated all logic into `processRawDocumentResponse`. It is possible that this feature will be removed in v10, but I wanted to make sure that missing processing of `raw_document` wouldn't be a blocker for enabling feature toggle in the future. Also it was easier and faster for me to implement this now, then returning back to it. 

`processRawDocumentResponse` processes response in the same way as it is currently done on frontend. 

When running test data, there might be a slight difference in:

-  ordering (results from backend are alphabetically ordered - this is done automatically while JSON.Marshall. From [docs](https://pkg.go.dev/encoding/json#Marshal):  `...map's key type must either be a string, an integer type, or implement encoding.TextMarshaler. The map keys are sorted`. This shouldn't be an issue at all, as the order of fields in json is not used anywhere.
- In json field (only field),  `@timestamp` is array with value when processed trough backend (compared to just value when processed trough frontend). This is because when we execute query trough backend, we receive `hit["fields"]` which has `@timestamp` with array with timestamp, compared to request trough frontend, where `hit["fields"]` is not present. 

This is because on backend, we are purposefully adding `time field` to ` b.AddDocValueField(defaultTimeField)` https://github.com/grafana/grafana/blob/main/pkg/tsdb/elasticsearch/time_series_query.go#L350 which means, that we are setting it as custom prop in our request to ES. However, this is missing in frontend implementation.

As we have processing for `fields` and we have method `AddDocValueField`, I suspect that not adding `time field` as `value field` is mistake and therefore the change is positive. 

FYI More on fields https://www.elastic.co/guide/en/elasticsearch/reference/current/search-fields.html. 

<img width="1850" alt="image" src="https://user-images.githubusercontent.com/30407135/221602544-736bfc97-d4c3-41c6-a68b-ed0371882b03.png">


**Special notes for your reviewer**:

To test this
1. Run ElasticSearch data source by `make devenv sources=elastic`
2. Use `elasticsearchBackendMigration` feature toggle to toggle if query is run trough frontend/backend
3. Run `Raw document` query
4. Results run trough frontend and backend should be same (with exception of @timestamp and ordering).
